collectors/metrics/cmd/metrics-collector/main.go

```diff
@@ -554,68 +537,241 @@ func initConfig(o *Options) (error, *forwarder.Config) {
 	transformer.With(metricfamily.TransformerFunc(metricfamily.PackMetrics))
 	transformer.With(metricfamily.TransformerFunc(metricfamily.SortMetrics))
 
-	isHypershift, err := metricfamily.CheckCRDExist(o.Logger)
-	if err != nil {
-		return err, nil
+	// TODO(saswatamcode): Kill this feature.
+	// This is too messy of an approach, to get hypershift specific labels into metrics we send.
+	// There is much better way to do this, with relabel configs.
+	// A collection agent shouldn't be calling out to Kube API server just to add labels.
+	if !o.DisableHyperShift {
+		isHypershift, err := metricfamily.CheckCRDExist(o.Logger)
+		if err != nil {
+			return nil, err
+		}
+		if isHypershift {
+			config, err := clientcmd.BuildConfigFromFlags("", "")
+			if err != nil {
+				return nil, errors.New("failed to create the kube config for hypershiftv1")
+			}
+			s := scheme.Scheme
+			if err := hyperv1.AddToScheme(s); err != nil {
+				return nil, errors.New("failed to add observabilityaddon into scheme")
+			}
+			hClient, err := client.New(config, client.Options{Scheme: s})
+			if err != nil {
+				return nil, errors.New("failed to create the kube client")
+			}
+
+			hyperTransformer, err := metricfamily.NewHypershiftTransformer(hClient, o.Logger, o.Labels)
+			if err != nil {
+				return nil, err
+			}
+			transformer.WithFunc(func() metricfamily.Transformer {
+				return hyperTransformer
+			})
+		}
 	}
-	if isHypershift {
-		hyperTransformer, err := metricfamily.NewHypershiftTransformer(o.Logger, nil, o.Labels)
+
+	// Configure matchers.
+	matchers := o.Matchers
+	if len(o.MatcherFile) > 0 {
+		data, err := os.ReadFile(o.MatcherFile)
 		if err != nil {
-			return err, nil
+			return nil, fmt.Errorf("unable to read match-file: %w", err)
+		}
+		matchers = append(matchers, strings.Split(string(data), "\n")...)
+	}
+	for i := 0; i < len(matchers); {
+		s := strings.TrimSpace(matchers[i])
+		if len(s) == 0 {
+			matchers = append(matchers[:i], matchers[i+1:]...)
+			continue
+		}
+		matchers[i] = s
+		i++
+	}
+
+	var statusClient client.Client
+
+	// TODO(saswatamcode): Evaluate better way for status reporting.
+	// Currently, it reports status for every 3 errors in forward request. This is too much of an overhead
```

Comment:
I'm not sure there is much to win here, given that it will be replaced by MCOA. In my view, it's not worth the effort. It's good enough.

Reply:
Hmm, MCOA would still use metrics-collector, right? Or are you saying we won't be reporting status from here in MCOA anymore?

Reply:
The metrics collector should be replaced by the prometheusAgent. Nothing is set in stone though.

Reply:
> The metrics collector should be replaced by the prometheusAgent. Nothing is set in stone though.

Yup me and @moadz discussed this, but I don't think it would be feasible yet, as you can't downsample, so would end up sending all raw 30s scrape interval data

Reply:
Ah yeah in that case it is, but still promagent would end up doing single large /federate call